### DISCUSSION TITLE
MetaMask Snaps Bogota Guide

### DISCUSSION THREAD

**Discussion by ritave** [MetaMask/snaps#821] [Announcements]
# MetaMask Snaps Bogota Guide

We have some features specifically for Devcon 2022. Here’s how you can get started using them.

## Installing Flask

1. Download [MetaMask Flask Bogota Build](https://output.circle-artifacts.com/output/job/ac54c30a-c43c-4004-9998-219d3b298fce/artifacts/0/builds-flask/metamask-flask-chrome-10.20.0-flask.0.zip)
2. Unzip the build into a folder
3. Allow using unpacked Chrome extensions
   1. Go to the Extensions page by entering `chrome://extensions` in a new tab.
      - Alternatively, click on the Extensions menu puzzle button and select **Manage Extensions** at the bottom of the menu.
      - Or, click the Chrome menu, hover over **More Tools,** then select **Extensions**.
   2. Enable Developer Mode by clicking the toggle switch next to **Developer mode**.
4. Click the **Load unpacked** button and select the unziped folder
   ![https://wd.imgix.net/image/BhuKGJaIeLNPW9ehns59NfwqKxF2/BzVElZpUtNE4dueVPSp3.png?auto=format](https://wd.imgix.net/image/BhuKGJaIeLNPW9ehns59NfwqKxF2/BzVElZpUtNE4dueVPSp3.png?auto=format)
5. Set up MetaMask Flask as usual
6. **Keep in mind: You can't have normal MetaMask and MetaMask Flask installed at the same time.**

### Keyrings

![Container](https://user-images.githubusercontent.com/1614945/194188170-c96a730b-ab12-4154-8c0d-63148bd99367.png)

> [Full specification](https://github.com/MetaMask/SIPs/blob/main/SIPS/sip-2.md)

> Bogota support:
>
> - [x] MultiChain provider
> - [x] Transparent dapp <=> snap communication
> - [x] User connection approval UI
> - [ ] Events support
> - [ ] MetaMask UI integration

Keyrings allow you integrate custom blockchains and protocols in a way that's transparent both to the user and the dapp.

##### Manifest

Extend the manifest with support for Keyrings. The `eip155` namespace means that the chain is EVM-compatible. The `eip155:5` chain is the Görli testnet.

```json
{
  "initialPermissions": {
    "snap_confirm": {},
    "snap_manageState": {},
    "endowment:keyring": {
      "namespaces": {
        "eip155": {
          "methods": [
            "eth_accounts",
            "eth_sendTransaction",
            "gnosis_watchSafe",
            "gnosis_createSafe"
          ],
          "chains": {
            "id": "eip155:5",
            "name": "Ethereum (Goerli)"
          }
        }
      }
    }
  }
}
```

##### Dapp

Connect to the wallet using the new MultiChainProvider. It seamlessly matches the dapp to a snap that implements the required behavior.

```javascript
import { MultiChainProvider } from "@metamask/multichain-provider";

const provider = new MultiChainProvider();
const { approval } = await provider.connect({
  requiredNamespaces: {
    eip155: {
      chains: ["eip155:5"],
      methods: [
        "eth_accounts",
        "eth_sendTransaction",
        "gnosis_watchSafe",
        "gnosis_createSafe",
      ],
    },
  },
});
const session = await approval();

const accounts = await provider.request({ method: "eth_accounts" });
```

##### Snap

Export a new object that implements a connection to the Keyring API.

```javascript
exports.keyring = {
  handleRequest: async ({ request }) => {
    switch (request.method) {
      case "eth_accounts":
        return getAccounts();
      case "gnosis_watchSafe":
        return watchSafe(request.params[0]);
      /* ... */
    }
  },
};
```

### TX Insights

<img width="472" alt="Screenshot 2022-10-05 at 20 00 38" src="https://user-images.githubusercontent.com/1614945/194190805-f40bb9c0-da26-4845-b3a1-2619e0f39a0a.png">

> [Full specification](https://github.com/MetaMask/SIPs/blob/main/SIPS/sip-3.md)

> Bogota support
>
> - [x] Transaction insights for contract interactions
> - [ ] Insights for account to account transfers

Transaction insights allows you to add a new tab to the MetaMask pre-transaction screen that shows additional information about the user's transaction.

##### Manifest

Extend your manifest to notify MetaMask that your snap supports insights.

```json
{
  "initialPermissions": {
    "endowment:transaction-insight": {}
  }
}
```

##### Snap

Expose a handler that will be called when the user views your snap's tab in the pre-transaction confirmation window. Return a key-value `insights` object which will be displayed in your tab.

```javascript
exports.onTransaction = async ({ transaction }) => {
  return {
    insights: { score: 42, "Is contract verified on Etherscan?": "Yes" },
  };
};
```

**Comment by livingrockrises** [comment]
Hello, I am getting below error from 'wallet_getSnaps'

A handler for SnapController:getPermittedSnaps has not been registered

Can you please advise

**Reply by ritave** [reply]
We're taking a look and trying to get a reproduction - will update you when we can reproduce and fix the build

**Reply by hmalik88** [reply]
fix is incoming...

**Reply by hmalik88** [reply]
link updated!

**Reply by livingrockrises** [reply]
I'm actually getting {'message': 'not found'} 
https://output.circle-artifacts.com/output/job/a8d828e6-1d55-452e-ae02-d9c0b23c3140/artifacts/0/builds-flask/metamask-flask-chrome-10.20.0.zip

**Reply by JonWong203** [reply]
im getting same problem with link^

**Reply by aalimsahin** [reply]
I am also getting same error too now

**Reply by hmalik88** [reply]
mind posting a screenshot of your console?

**Reply by ritave** [reply]
In the meantime, you can use the basic template snap repositories. They avoid the problem.

[Javascript](https://github.com/MetaMask/template-snap)
[Typescript](https://github.com/MetaMask/template-typescript-snap)

**Reply by FrederikBolding** [reply]
Correct build link: https://output.circle-artifacts.com/output/job/a8d828e6-1d55-452e-ae02-d9c0b23c3140/artifacts/0/builds-flask/metamask-flask-chrome-10.20.0-flask.0.zip

**Reply by aalimsahin** [reply]
<img width="489" alt="image" src="https://user-images.githubusercontent.com/73793382/194629392-0e932ad7-fd74-474f-b9eb-4949fe4ddcc5.png">

**Reply by livingrockrises** [reply]
it now works for me, thank you!

**Reply by franz101** [reply]
works here too

**Comment by tomarsachin2271** [comment]
Using snaps, can we change the account displayed on metamask UI?

**Reply by ritave** [reply]
No

**Reply by tomarsachin2271** [reply]
So "snap_confirm" is the only way to show eg., BLS public key or any snap specific data to the user? Is something like this going to be added in the future which will allow us to change metamask UI components also?

**Reply by ritave** [reply]
Yes! We're researching a declarative UI, similar to [Slack's Block Kit UI](https://api.slack.com/block-kit), where your snap can return a JSON object to show a ui to a user.

It's not implemented yet, but here's an [API I was thinking about](https://gist.github.com/ritave/936013d84a0111ae27046e940915c5cc).

**Comment by tomarsachin2271** [comment]
Also how can we get access to Metamask EOA account private key if possible?? I know we can request metamask to sign messages but if we are able to request permission to access the private key stored in meatmask storage, we'd be able to sign messages without showing any popup to the user. 

I came across this snap https://github.com/MetaMask/snap-examples/blob/main/examples/bls-signer/index.js

where its requesting the "snap_getAppKey" method to get metamask EOA private key but this method is not working in latest build.

**Reply by ritave** [reply]
You can request the [private keys for any coin type](https://docs.metamask.io/guide/snaps-rpc-api.html#snap-getbip44entropy). Ethereum's coin type is 60, so that would be `snap_getBip44Entropy_60` permission

**Comment by tomarsachin2271** [comment]
How to let snap code know that metamask EOA has been changed?
I'm trying 

```
wallet.on("accountsChanged", (accounts) => {
    console.log(accounts);
})
```

But it doesn't seem to be working. Any solution to this?

**Reply by ritave** [reply]
You probably need to allow the snap to get access to `eth_accounts` using `eth_requestAccounts` - https://docs.metamask.io/guide/rpc-api.html#restricted-methods

If that doesn't work you could probably just poll every few seconds using `eth_accounts`

**Reply by tomarsachin2271** [reply]
Its working. Thanks

**Comment by gnardini** [comment]
I'm building a smart contract wallet so very interested in the keyrings feature.

Is there any public example you can point me to? I'm reading the docs and have a vague understanding on how to use it but I generally learn a lot faster by seeing working examples.

Thanks a lot!

**Reply by ritave** [reply]
Sadly we didn’t have time make examples. I’ll be happy to walk you through tomorrow. @ritave on Discord

**Reply by gnardini** [reply]
Thanks! I'll ping you tomorrow with some specific questions I'm stuck with 😅

**Reply by gnardini** [reply]
Hey, can't seem to find you on Discord :( my username is Gonza#2410 would you mind DMing me? :)

**Comment by tomarsachin2271** [comment]
I've configured this permission to get network access `endowment:network-access` but somehow i'm not able to do rpc calls on goerli network...
is there anything else i need to do to enable network access?
i'm using normal ethers provider to read values from smart contract

**Reply by Montoya** [reply]
Hi, sorry I missed this before, did you manage to solve this @tomarsachin2271?

**Comment by 0xalank** [comment]
Hey! I'm seeing`Error: Keyring does not expose getAccounts` on this build when using the multichain-provider. Is there something I'm missing?

**Reply by 0xalank** [reply]
nvm found that I wasn't adding the getAccounts method like so:
```
class Keyring {
  async getAccounts() {
    return ['eip155:1:foo'];
  }
  async handleRequest({request}) {
    switch(request.method){
      case 'eth_accounts':
        return this.getAccounts();
    }
  }
}
module.exports.keyring = new Keyring();

```

**Reply by ritave** [reply]
Hey @alanorwick!

You need to expose `keyring.getAccounts()` method as well. It's used during connection flow to ask users which accounts to allow the dApp to access.

Read more about the [method in the specification](https://github.com/MetaMask/SIPs/blob/main/SIPS/sip-2.md#snap).

Here's an example.

```typescript
const CAIP_2_CHAIN_ID = 'eip155:1';
async function getAccounts() {
  /*
   There's currently a chicken and egg problem, where MetaMask requires at least 1 account
   to show before allowing a dApp to connect, but there's no way to create accounts using
   the UI. We solve this problem by returning a dummy account for now.
   */
  const state = await getState() ?? { accounts: [`${CAIP_2_CHAIN_ID}:0x9b2055d370f73ec7d8a03e965129118dc8f5bf83`] };
  return state.accounts;
}

function ethToCaip10(ethAddress: string) {
  return `${CAIP_2_CHAIN_ID}:${ethAddress}`;
}

async function sign(account: string, message: string) {
  /* ... */
}

function handleRequest({ request }) {
  switch (request.method) {
    case 'eth_sign':
       const [account, message] = request.params;
       const accounts = await getAccounts();
       if (!accounts.includes(ethToCaip10(account)) {
         throw new Error("Invalid account");
       }
       return sign(account, message);
  }
}

exports.keyring = {
  getAccounts,
  handleRequest
};
```

**Reply by 0xalank** [reply]
Got it all working! Thank you for the insight. Any advice on how to get that signing part done that you described above? Will that require `snap_getBip44Entropy`? Apologies if I'm getting too ahead of development here 😄

**Reply by ritave** [reply]
Generally in the above example you would generate the accounts using `snap_getBip44Entropy` and `@metamask/key-tree` instead of hardcoding. And then since account generation should yield the same results using the same entropy, you would also use the same entropy to calculate signatures as well.

To implement sign you would use `snap_getBip44Entropy`, use `@metamask/key-tree` to generate private keys for the requested account, and then use `@ethereumjs/utils` [sign methods](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/util/src/signature.ts) to calculate the signature.
